lib/rubygems/gemcutter_utilities.rb

```diff
@@ -250,11 +251,28 @@ def request_with_otp(method, uri, &block)
     end
   end
 
-  def ask_otp
-    say "You have enabled multi-factor authentication. Please enter OTP code."
+  def ask_otp(credentials)
+    webauthn_url = webauthn_verification_url(credentials)
+    if webauthn_url
+      say "You have enabled multi-factor authentication. Please enter OTP code from your security device by visiting #{webauthn_url}."
+    else
+      say "You have enabled multi-factor authentication. Please enter OTP code."
+    end
+
     options[:otp] = ask "Code: "
   end
 
+  def webauthn_verification_url(credentials)
+    response = rubygems_api_request(:post, "api/v1/webauthn_verification") do |request|
+      if credentials
+        request.basic_auth credentials[:email], credentials[:password]
+      else
+        request.add_field "Authorization", api_key
+      end
+    end
+    response.is_a?(Net::HTTPSuccess) ? response.body : nil
```

Comment:
For hosts that do not have webauthn / "api/v1/webauthn_verification" implemented yet, I'm guessing that there's not going to be a change in behaviour since this method would return nil.